Vulkan-Pipeline-Assistant/spvvectorwidget.cpp

```diff
@@ -0,0 +1,75 @@
+#include "spvvectorwidget.h"
+
+#include <QLineEdit>
+#include <QLabel>
+#include <QLayout>
+#include <QPushButton>
+#include <QCoreApplication>
+#include <QValidator>
+
+#include "spirvresource.h"
+
+using namespace vpa;
+
+const QString SpvVectorWidget::VecStrings[4] = { "float", "vec2", "vec3", "vec4" };
+const QString SpvVectorWidget::ComponentStrings[4] = { "x", "y", "z", "w" };
+
+SpvVectorWidget::SpvVectorWidget(SpvVectorType* type, QWidget* parent)
+    : SpvWidget(parent), m_type(type) {
+    QVBoxLayout* layout = new QVBoxLayout(this);
+    layout->setAlignment(Qt::AlignTop);
+
+    m_infoGroup = new QWidget(parent);
+    m_infoGroup->setLayout(new QHBoxLayout(m_infoGroup));
+    m_infoGroup->layout()->addWidget(new QLabel(VecStrings[(m_type->length - 1)], parent));
+    QPushButton* normBtn = new QPushButton("Normalize", m_infoGroup);
+    m_infoGroup->layout()->addWidget(normBtn);
+    connect(normBtn, SIGNAL(pressed()), this, SLOT(HandleNormalize()));
+    m_infoGroup->layout()->setAlignment(Qt::AlignTop);
+
+    m_inputsGroup = new QWidget(parent);
+    QGridLayout* inputsGroupLayout = new QGridLayout(m_inputsGroup);
```

Comment:
Does this memory need to be freed locally?

Reply:
No it doesn't, Qt adds it to its own memory management system as I understand it. When the parent is destroyed, all children will be. Ultimately, the window is destroyed and it is at the top of the parent chain.